File: Data Structures & Algorithms/clone-graph/submission-1.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        # dfs og graph
        nodes = dict()
        visited = set()
        def dfs(node):
            if node is None:
                return

            if node.val in visited:
                return
            else:
                visited.add(node.val)

            if node.val in nodes:
                newNode = nodes[node.val]
            else:
                newNode = Node(node.val)
                nodes[node.val] = newNode
            
            for nb in node.neighbors:
                if nb.val in nodes:
                    newNb = nodes[nb.val]
                else:
                    newNb = Node(nb.val)
                    nodes[nb.val] = newNb

                newNode.neighbors.append(newNb)
                dfs(nb)
        
        dfs(node)
        
        if node is None:
            return None
        return nodes[node.val]

File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
class Solution {
public:
    array<array<int, 2>, 4> dirs{{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}};
    int dfs(int i, int j, vector<vector<bool>>& visited, const vector<vector<char>>& grid) {
        if (i < 0 || i == grid.size() || j < 0 || j == grid[0].size()) return 0;
        if (visited[i][j] || grid[i][j] == '0') return 0;
        visited[i][j] = true;
        for (const auto[di, dj] : dirs) {
            dfs(i + di, j + dj, visited, grid);
        }
        return 1;
    }
    int numIslands(vector<vector<char>>& grid) {
        vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size(), false));
        auto islands{0};
        for (auto i{0}; i < grid.size(); ++i) {
            for (auto j{0}; j < grid[0].size(); ++j) {
                islands += dfs(i, j, visited, grid);
            }
        }
        return islands;
    }
};
